### PR TITLE
Fix RT-DETR inference with float16 and bfloat16

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?

Fix positional embeddings and anchor data types for RT-DETR.

Running the model in `float16` or `bfloat16` with the following code crashes with the error:

```python
import torch
import requests

from PIL import Image
from transformers import RTDetrForObjectDetection, RTDetrImageProcessor

device = "cuda"
dtype = torch.float16

url = 'http://images.cocodataset.org/val2017/000000039769.jpg' 
image = Image.open(requests.get(url, stream=True).raw)

image_processor = RTDetrImageProcessor.from_pretrained("PekingU/rtdetr_r50vd")
model = RTDetrForObjectDetection.from_pretrained("PekingU/rtdetr_r50vd", attn_implementation="eager", torch_dtype=dtype, device_map=device)

inputs = image_processor(images=image, return_tensors="pt").to(device)
pixel_values = inputs.pixel_values.to(device).to(dtype)

with torch.no_grad():
    outputs = model(pixel_values)

assert outputs.logits.dtype == dtype

results = image_processor.post_process_object_detection(outputs, target_sizes=torch.tensor([image.size[::-1]]), threshold=0.3)

for result in results:
    for score, label_id, box in zip(result["scores"], result["labels"], result["boxes"]):
        score, label = score.item(), label_id.item()
        box = [round(i, 2) for i in box.tolist()]
        print(f"{model.config.id2label[label]}: {score:.2f} {box}")

```
Error:
```
RuntimeError: expected mat1 and mat2 to have the same dtype, but got: float != c10::Half
```

One approach to fix is to use `torch.autocast()`, but it could also be fixed with proper dtypes in the modeling file.

Steps:

1. Add test that fails with float16/bfloat16 dtypes
2. Add fix

## Before submitting
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qubvel** [comment]
[Added test failed](https://github.com/huggingface/transformers/actions/runs/9680176397/job/26707756431?pr=31639) for float16 and bfloat16 as expected:

Screenshot 2024-06-26 at 14 14 00

**Comment by qubvel** [comment]
Tests are green with the fix

File: src/transformers/models/rt_detr/modeling_rt_detr.py [hunk 1]
```diff
@@ -1357,9 +1357,13 @@ def forward(
...
                 if self.training or self.eval_size is None:
-                    pos_embed = self.build_2d_sincos_position_embedding(
-                        width, height, self.encoder_hidden_dim, self.positional_encoding_temperature
-                    ).to(src_flatten.device)
+                    pos_embed = (
+                        self.build_2d_sincos_position_embedding(
+                            width, height, self.encoder_hidden_dim, self.positional_encoding_temperature
+                        )
+                        .to(src_flatten.device)
+                        .to(src_flatten.dtype)
```
Comment (amyeroberts) [inline src/transformers/models/rt_detr/modeling_rt_detr.py:1365]: We can pass dtype and device in the same `to` call

```suggestion
                        .to(src_flatten.device, src_flatten.dtype)
```

File: src/transformers/models/rt_detr/modeling_rt_detr.py [hunk 2]
```diff
@@ -1801,12 +1805,13 @@ def forward(
...
         device = source_flatten.device
+        dtype = source_flatten.dtype
 
         # prepare input for decoder
         if self.training or self.config.anchor_image_size is None:
-            anchors, valid_mask = self.generate_anchors(spatial_shapes, device=device)
+            anchors, valid_mask = self.generate_anchors(spatial_shapes, device=device, dtype=dtype)
         else:
-            anchors, valid_mask = self.anchors.to(device), self.valid_mask.to(device)
+            anchors, valid_mask = self.anchors.to(device).to(dtype), self.valid_mask.to(device).to(dtype)
```
Comment (amyeroberts) [inline src/transformers/models/rt_detr/modeling_rt_detr.py:1814]:
```suggestion
            anchors, valid_mask = self.anchors.to(device, dtype), self.valid_mask.to(device, dtype)
```

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!

**Comment by qubvel** [comment]
Thanks for the review!

**Pull request merged**